crypto/ec/ecp_nistp384.c

```diff
@@ -0,0 +1,1935 @@
+           + ((uint128_t) in[2]) * inx2[5]
+           + ((uint128_t) in[3]) * inx2[4];
+
+    out[8] = ((uint128_t) in[2]) * inx2[6]
+           + ((uint128_t) in[3]) * inx2[5]
+           + ((uint128_t) in[4]) * in[4];
+
+    out[9] = ((uint128_t) in[3]) * inx2[6]
+           + ((uint128_t) in[4]) * inx2[5];
+
+    out[10] = ((uint128_t) in[4]) * inx2[6]
+            + ((uint128_t) in[5]) * in[5];
+
+    out[11] = ((uint128_t) in[5]) * inx2[6];
+
+    out[12] = ((uint128_t) in[6]) * in[6];
+}
+
+static void felem_mul_ref(widefelem out, const felem in1, const felem in2)
+{
+    out[0] = ((uint128_t) in1[0]) * in2[0];
+
+    out[1] = ((uint128_t) in1[0]) * in2[1]
+           + ((uint128_t) in1[1]) * in2[0];
+
+    out[2] = ((uint128_t) in1[0]) * in2[2]
+           + ((uint128_t) in1[1]) * in2[1]
+           + ((uint128_t) in1[2]) * in2[0];
+
+    out[3] = ((uint128_t) in1[0]) * in2[3]
+           + ((uint128_t) in1[1]) * in2[2]
+           + ((uint128_t) in1[2]) * in2[1]
+           + ((uint128_t) in1[3]) * in2[0];
+
+    out[4] = ((uint128_t) in1[0]) * in2[4]
+           + ((uint128_t) in1[1]) * in2[3]
+           + ((uint128_t) in1[2]) * in2[2]
+           + ((uint128_t) in1[3]) * in2[1]
+           + ((uint128_t) in1[4]) * in2[0];
+
+    out[5] = ((uint128_t) in1[0]) * in2[5]
+           + ((uint128_t) in1[1]) * in2[4]
+           + ((uint128_t) in1[2]) * in2[3]
+           + ((uint128_t) in1[3]) * in2[2]
+           + ((uint128_t) in1[4]) * in2[1]
+           + ((uint128_t) in1[5]) * in2[0];
+
+    out[6] = ((uint128_t) in1[0]) * in2[6]
+           + ((uint128_t) in1[1]) * in2[5]
+           + ((uint128_t) in1[2]) * in2[4]
+           + ((uint128_t) in1[3]) * in2[3]
+           + ((uint128_t) in1[4]) * in2[2]
+           + ((uint128_t) in1[5]) * in2[1]
+           + ((uint128_t) in1[6]) * in2[0];
+
+    out[7] = ((uint128_t) in1[1]) * in2[6]
+           + ((uint128_t) in1[2]) * in2[5]
+           + ((uint128_t) in1[3]) * in2[4]
+           + ((uint128_t) in1[4]) * in2[3]
+           + ((uint128_t) in1[5]) * in2[2]
+           + ((uint128_t) in1[6]) * in2[1];
+
+    out[8] = ((uint128_t) in1[2]) * in2[6]
+           + ((uint128_t) in1[3]) * in2[5]
+           + ((uint128_t) in1[4]) * in2[4]
+           + ((uint128_t) in1[5]) * in2[3]
+           + ((uint128_t) in1[6]) * in2[2];
+
+    out[9] = ((uint128_t) in1[3]) * in2[6]
+           + ((uint128_t) in1[4]) * in2[5]
+           + ((uint128_t) in1[5]) * in2[4]
+           + ((uint128_t) in1[6]) * in2[3];
+
+    out[10] = ((uint128_t) in1[4]) * in2[6]
+            + ((uint128_t) in1[5]) * in2[5]
+            + ((uint128_t) in1[6]) * in2[4];
+
+    out[11] = ((uint128_t) in1[5]) * in2[6]
+            + ((uint128_t) in1[6]) * in2[5];
+
+    out[12] = ((uint128_t) in1[6]) * in2[6];
+}
+
+/*-
+ * Reduce thirteen 128-bit coefficients to seven 64-bit coefficients.
+ * in[i] < 2^128 - 2^125
+ * out[i] < 2^56 for i < 6,
+ * out[6] <= 2^48
+ */
+static void felem_reduce(felem out, const widefelem in)
```

Comment:
Is there any chance of explaining how you come up with all of the numbers?

Reply:
Are we talking about the inequalities or about the program constants?

When it comes to the subtraction constants, they are chosen as power of two multiples of $p_{384}$ that are sufficiently large so that each limb can be sufficiently large. This is achieved through a telescopic sum. You'll notice that we subtract $2^{68}$ from a high limb whenever we add $2^{124}$ to a lower limb. Since $2^{124} = 2^{56} \cdot 2^{68}$ these two alterations cancel out, but lead to each limb being roughly $2^{124}$ in size. This is helpful, as adding a multiple of our modulus like so allows us to avoid underflow when subtracting limbs, which is a requirement for correctness.